### PR TITLE
Ginkgo: Migrated tests/21-ct-clean-up.sh

### PR DESCRIPTION
**Pull request by eloycoto**
Migrated CT clean up test to Ginkgo. Related to the task #1916

This code does not migrate all the file, the Envoy proxy test is not in place. Need to work on the build first.

The task is not closed yet, so a new one is not needed. 

Signed-off-by: Eloy Coto <eloy.coto@gmail.com>

### CODE REVIEW

File: test/helpers/cilium.go
Comment (ianvernon) [inline test/helpers/cilium.go:332]: Does this command have an option for JSON output? I want to move away from using `grep` and `awk` as it makes the output we parse dependent upon ordering as opposed to a key-value mapping which JSON provides.

File: test/helpers/cilium.go
Comment (ianvernon) [inline test/helpers/cilium.go:331]: What is the motivation for this change to this function? What does ` s.ExecCilium(fmt.Sprintf("policy delete %s", id))` not provide?

File: test/helpers/cilium.go
Comment (ianvernon) [inline test/helpers/cilium.go]: receices --> receives?

File: test/helpers/cilium.go
Comment (ianvernon) [inline test/helpers/cilium.go]: imports the policy to Cilium.

File: test/helpers/cilium.go
Comment (ianvernon) [inline test/helpers/cilium.go]: or if the policy cannot be imported

File: test/runtime/ct.go
Comment (ianvernon) [inline test/runtime/ct.go]: Move this import to be before the one on line 11.

File: test/runtime/ct.go
Comment (ianvernon) [inline test/runtime/ct.go]: `ContainerCreate` returns a `*CmdRes` yet the resultant `*CmdRest` is ignored here. Why is it ignored?

File: test/runtime/ct.go
Comment (ianvernon) [inline test/runtime/ct.go]: Either explicitly ignore the output of functions using `_` and add a comment explaining why you are ignoring the output, or check whether the error returned from this function is nil or not. I am against silently ignoring the output of helper functions, as it's easy to miss while reviewing, and ignoring the output of functions might result in more difficult debugging of test failures.

File: test/runtime/ct.go
Comment (ianvernon) [inline test/runtime/ct.go]: Why is the resultant `*CmdRes` ignored here?

File: test/runtime/ct.go
Comment (ianvernon) [inline test/runtime/ct.go]: Why is the resultant `*CmdRes` ignored here?

File: test/runtime/ct.go
Comment (ianvernon) [inline test/runtime/ct.go]: Why is the resultant `*CmdRes` ignored here?

File: test/runtime/ct.go
Comment (ianvernon) [inline test/runtime/ct.go]: Please add comments explaining what these strings represent. Since types aren't being used here, it's important to document what these strings represent.

File: test/runtime/ct.go
Comment (ianvernon) [inline test/runtime/ct.go:137]: Instead of using grep, can JSON parsing be done for this command?

File: test/runtime/ct.go
Comment (ianvernon) [inline test/runtime/ct.go]: Make `"Name"` into a constant and utilize it throughout this test instead of hardcoding it.

File: test/runtime/ct.go
Comment (ianvernon) [inline test/runtime/ct.go]: CTables --> CT tables

**Review by ianvernon — CHANGES_REQUESTED** [review]

File: test/runtime/ct.go
Comment (eloycoto) [reply test/runtime/ct.go]: The thing is that NetworkCreate sometimes failed that it's already created. Maybe added a function like now (NetworkCreateIfNotExists)?

**Comment by eloycoto** [comment]
@ianvernon The commands don't have JSON support yet, so not awk and grep is needed.  :-( 

Regards

File: test/runtime/ct.go
Comment (ianvernon) [reply test/runtime/ct.go]: +1, that function would be helpful to have.

**Comment by ianvernon** [comment]
@eloycoto thanks for confirming.

**Review by ianvernon — APPROVED** [review]
LGTM, but I want approval from @aanm as he is the author of the original test before merging.

File: test/runtime/ct.go
Comment (aanm) [inline test/runtime/ct.go:15]: Missing header (2018)

File: test/runtime/ct.go
Comment (aanm) [inline test/runtime/ct.go]: Can't you use `sync.Once`?

File: test/runtime/ct.go
Comment (aanm) [inline test/runtime/ct.go:99]: Out of curiosity, in case of failure, where are the already started containers going to be deleted?

File: test/runtime/ct.go
Comment (aanm) [inline test/runtime/ct.go:102]: Why was this container started on a different way than the previous 4?

File: test/runtime/ct.go
Comment (aanm) [inline test/runtime/ct.go:107]: Why was this container started on a different way than the previous 4?

File: test/runtime/ct.go
Comment (aanm) [inline test/runtime/ct.go]: Are you sure the test was checking for `Equal` and not `LessThan`?

File: test/runtime/ct.go
Comment (aanm) [inline test/runtime/ct.go]: Are you sure the test was checking for `Equal` and not `LessThan`?

File: test/runtime/ct.go
Comment (aanm) [inline test/runtime/ct.go]: I don't understand the message `"CT Tables are not correct for %s"`, what did you intend with this?

File: test/runtime/ct.go
Comment (aanm) [inline test/runtime/ct.go]: s/`reuising`/`reusing`/

File: test/runtime/ct.go
Comment (aanm) [inline test/runtime/ct.go]: Are you sure the test was checking for `Equal` and not `LessThan`?

File: test/runtime/manifests/ct-clean-up-nc.py
Comment (aanm) [inline test/runtime/manifests/ct-clean-up-nc.py]: s/`21-ct-clean-up-nc.py`/`ct-clean-up-nc.py`

**Review by aanm — CHANGES_REQUESTED** [review]

File: test/runtime/ct.go
Comment (eloycoto) [reply test/runtime/ct.go:99]: Yes, `defer containers(helpers.Delete)`  ;-)

File: test/runtime/ct.go
Comment (eloycoto) [reply test/runtime/ct.go:102]: The main reason of that is the `vm.ContainerCreate()` does not support volumes or commands.

File: test/runtime/ct.go
Comment (eloycoto) [reply test/runtime/ct.go:107]: The main reason of that is the `vm.ContainerCreate()` does not support volumes or commands.

File: test/runtime/ct.go
Comment (aanm) [inline test/runtime/ct.go]: What's the reason for this flush? It wasn't present in the old test.

File: test/runtime/ct.go
Comment (aanm) [inline test/runtime/ct.go]: It seems the test is not complete, it's lacking all tests after line 422 from `21-ct-clean-up.sh` (`# Install a L4 policy and confirm connectivity`)

**Review by aanm — CHANGES_REQUESTED** [review]

File: test/runtime/ct.go
Comment (aanm) [inline test/runtime/ct.go]: This port should be `11111`

**Review by aanm — CHANGES_REQUESTED** [review]
@eloycoto I believe this is still wip because the conversion of the old test is not 100% done, right?

**Comment by eloycoto** [comment]
@aanm this should be ok now ;-)

**Comment by ianvernon** [comment]
test-me-please

**Comment by ianvernon** [comment]
test-ginkgo-all

**Comment by ianvernon** [comment]
@eloycoto can you rebase and re-trigger tests, including all Ginkgo tests?

**Comment by eloycoto** [comment]
test-me-please

**Comment by eloycoto** [comment]
test-all-ginkgo

**Comment by eloycoto** [comment]
test-all-ginkgo

**Comment by eloycoto** [comment]
test-all-ginkgo

**Comment by ianvernon** [comment]
@aanm please review

**Comment by aanm** [comment]
@ianvernon I've talked with @eloycoto. I would like to fix issue #2408 first before merging this PR

**Comment by ianvernon** [comment]
@aanm does #2408 also affect the bash-script-based framework?

**Review by aanm — APPROVED** [review]

**Pull request merged**
